File: SatinalmaPro.Shared/Helpers/KalemFirmaAtamaYardimcisi.cs
using SatinalmaPro.Shared.Models;

namespace SatinalmaPro.Shared.Helpers;

/// <summary>
/// Kalem miktarını birden fazla firmaya bölme — geriye uyumlu tek OnaylananTeklifId desteği.
/// </summary>
public static class KalemFirmaAtamaYardimcisi
{
    public const double Tolerans = 0.0001;

    public static bool OnayliMi(SatinalmaTalepKalemi kalem) =>
        EtkinAtamalar(kalem).Count > 0;

    /// <summary>
    /// FirmaAtamalari doluysa onu; değilse OnaylananTeklifId → tek tam-miktar atama.
    /// </summary>
    public static List<KalemFirmaAtamasi> EtkinAtamalar(SatinalmaTalepKalemi kalem)
    {
        var liste = kalem.FirmaAtamalari?
            .Where(a => a.TeklifId != Guid.Empty && a.Miktar > Tolerans)
            .OrderByDescending(a => a.Miktar)
            .ToList() ?? [];

        if (liste.Count > 0)
            return liste;

        if (kalem.OnaylananTeklifId is { } tid)
        {
            return
            [
                new KalemFirmaAtamasi
                {
                    TeklifId = tid,
                    Miktar = kalem.Miktar,
                    KabulEdilenMiktar = kalem.KabulEdilenMiktar,
                    SiparisTamamlandi = kalem.SiparisTamamlandi
                }
            ];
        }

        return [];
    }

    public static IEnumerable<Guid> OnayliTeklifIdleri(SatinalmaTalepKalemi kalem) =>
        EtkinAtamalar(kalem).Select(a => a.TeklifId).Distinct();

    public static void Dogrula(SatinalmaTalepKalemi kalem, IReadOnlyList<KalemFirmaAtamasi> atamalar)
    {
        if (atamalar.Count == 0)
            throw new InvalidOperationException($"«{kalem.Malzeme}» için en az bir firma ataması gerekli.");

        if (atamalar.Any(a => a.Miktar <= Tolerans))
            throw new InvalidOperationException($"«{kalem.Malzeme}» atama miktarları sıfırdan büyük olmalı.");

        if (atamalar.GroupBy(a => a.TeklifId).Any(g => g.Count() > 1))
            throw new InvalidOperationException($"«{kalem.Malzeme}» için aynı firma birden fazla kez seçilemez.");

        var toplam = atamalar.Sum(a => a.Miktar);
        if (Math.Abs(toplam - kalem.Miktar) > Tolerans)
            throw new InvalidOperationException(
                $"«{kalem.Malzeme}» atama toplamı ({toplam:N2}) talep miktarına ({kalem.Miktar:N2}) eşit olmalı.");
    }

    /// <summary>Atamaları yazar; OnaylananTeklifId = en büyük miktarlı firma; kalem kabul özetini senkronlar.</summary>
    public static void Uygula(SatinalmaTalepKalemi kalem, IEnumerable<KalemFirmaAtamasi> atamalar)
    {
        var liste = atamalar
            .Where(a => a.TeklifId != Guid.Empty && a.Miktar > Tolerans)
            .Select(a => new KalemFirmaAtamasi
            {
                TeklifId = a.TeklifId,
                Miktar = a.Miktar,
                KabulEdilenMiktar = a.KabulEdilenMiktar,
                SiparisTamamlandi = a.SiparisTamamlandi
            })
            .ToList();

        Dogrula(kalem, liste);
        kalem.FirmaAtamalari = liste;
        kalem.OnaylananTeklifId = liste
            .OrderByDescending(a => a.Miktar)
            .Select(a => (Guid?)a.TeklifId)
            .First();
        KabulOzetiniSenkronla(kalem);
    }

    public static void TekFirmayaAta(SatinalmaTalepKalemi kalem, Guid teklifId) =>
        Uygula(kalem, [new KalemFirmaAtamasi { TeklifId = teklifId, Miktar = kalem.Miktar }]);

    public static void Temizle(SatinalmaTalepKalemi kalem)
    {
        kalem.FirmaAtamalari = [];
        kalem.OnaylananTeklifId = null;
        kalem.KabulEdilenMiktar = 0;
        kalem.SiparisTamamlandi = false;
    }

    public static KalemFirmaAtamasi? AtamaBul(SatinalmaTalepKalemi kalem, Guid teklifId) =>
        EtkinAtamalar(kalem).FirstOrDefault(a => a.TeklifId == teklifId);

    public static void KabulEkle(SatinalmaTalepKalemi kalem, Guid teklifId, double miktar)
    {
        var atamalar = EtkinAtamalar(kalem).Select(a => new KalemFirmaAtamasi
        {
            TeklifId = a.TeklifId,
            Miktar = a.Miktar,
            KabulEdilenMiktar = a.KabulEdilenMiktar,
            SiparisTamamlandi = a.SiparisTamamlandi
        }).ToList();

        var atama = atamalar.FirstOrDefault(a => a.TeklifId == teklifId)
            ?? throw new InvalidOperationException("Firma ataması bulunamadı.");

        atama.KabulEdilenMiktar += miktar;
        if (atama.KabulEdilenMiktar > atama.Miktar + Tolerans)
            atama.Miktar = atama.KabulEdilenMiktar;
        if (atama.KabulEdilenMiktar >= atama.Miktar - Tolerans)
            atama.SiparisTamamlandi = true;

        kalem.FirmaAtamalari = atamalar;
        if (kalem.OnaylananTeklifId is null)
            kalem.OnaylananTeklifId = atamalar.OrderByDescending(a => a.Miktar).First().TeklifId;
        KabulOzetiniSenkronla(kalem);
    }

    public static void SevkiyatiTamamla(SatinalmaTalepKalemi kalem, Guid teklifId)
    {
        var atamalar = EtkinAtamalar(kalem).Select(a => new KalemFirmaAtamasi
        {
            TeklifId = a.TeklifId,
            Miktar = a.Miktar,
            KabulEdilenMiktar = a.KabulEdilenMiktar,
            SiparisTamamlandi = a.SiparisTamamlandi
        }).ToList();

        var atama = atamalar.FirstOrDefault(a => a.TeklifId == teklifId)
            ?? throw new InvalidOperationException("Firma ataması bulunamadı.");

        if (atama.KabulEdilenMiktar < atama.Miktar - Tolerans)
            atama.Miktar = atama.KabulEdilenMiktar;
        atama.SiparisTamamlandi = true;

        kalem.FirmaAtamalari = atamalar;
        KabulOzetiniSenkronla(kalem);
    }

    public static void KabulOzetiniSenkronla(SatinalmaTalepKalemi kalem)
    {
        var atamalar = kalem.FirmaAtamalari?
            .Where(a => a.TeklifId != Guid.Empty && a.Miktar > Tolerans)
            .ToList() ?? [];

        if (atamalar.Count == 0)
            return;

        kalem.KabulEdilenMiktar = atamalar.Sum(a => a.KabulEdilenMiktar);
        kalem.SiparisTamamlandi = atamalar.All(a =>
            a.SiparisTamamlandi || a.KabulEdilenMiktar >= a.Miktar - Tolerans);
        // Kalem toplam miktarı atama toplamına hizala (fazla teslim sonrası)
        var atamaToplam = atamalar.Sum(a => a.Miktar);
        if (atamaToplam > Tolerans && Math.Abs(atamaToplam - kalem.Miktar) > Tolerans)
            kalem.Miktar = atamaToplam;
    }

    public static string OzetMetni(
        SatinalmaTalepKalemi kalem,
        IEnumerable<SatinalmaTeklif> teklifler)
    {
        var atamalar = EtkinAtamalar(kalem);
        if (atamalar.Count == 0)
            return "Firma seçilmedi";

        return string.Join(" + ", atamalar.Select(a =>
        {
            var firma = teklifler.FirstOrDefault(t => t.Id == a.TeklifId)?.FirmaAdi ?? "?";
            return $"{a.Miktar:N2} {firma}";
        }));
    }

    /// <summary>
    /// Firma miktarını ayarlar; diğer firmaların payını kalan miktara oransal ölçekler.
    /// </summary>
    public static void FirmaMiktariniAyarla(SatinalmaTalepKalemi kalem, Guid teklifId, double miktar)
    {
        if (miktar <= Tolerans)
            throw new InvalidOperationException("Miktar sıfırdan büyük olmalı.");
        if (miktar > kalem.Miktar + Tolerans)
            throw new InvalidOperationException("Miktar talep miktarını aşamaz.");

        var diger = EtkinAtamalar(kalem)
            .Where(a => a.TeklifId != teklifId)
            .Select(a => new KalemFirmaAtamasi
            {
                TeklifId = a.TeklifId,
                Miktar = a.Miktar,
                KabulEdilenMiktar = a.KabulEdilenMiktar,
                SiparisTamamlandi = a.SiparisTamamlandi
            })
            .ToList();

        var kalan = kalem.Miktar - miktar;
        if (diger.Count == 0)
        {
            if (Math.Abs(miktar - kalem.Miktar) > Tolerans)
                throw new InvalidOperationException(
                    "Tek firmada miktar talep toplamına eşit olmalı. Başka firma ekleyin veya tüm miktarı seçin.");
            Uygula(kalem, [new KalemFirmaAtamasi { TeklifId = teklifId, Miktar = miktar }]);
            return;
        }

        var sonuc = new List<KalemFirmaAtamasi>();
        if (kalan > Tolerans)
        {
            var digerToplam = diger.Sum(a => a.Miktar);
            if (digerToplam <= Tolerans)
            {
                diger[0].Miktar = kalan;
                sonuc.Add(diger[0]);
            }
            else
            {
                foreach (var a in diger)
                {
                    a.Miktar = a.Miktar / digerToplam * kalan;
                    if (a.Miktar > Tolerans)
                        sonuc.Add(a);
                }

                var fark = kalan - sonuc.Sum(a => a.Miktar);
                if (sonuc.Count > 0)
                    sonuc[0].Miktar += fark;
            }
        }

        sonuc.Add(new KalemFirmaAtamasi { TeklifId = teklifId, Miktar = miktar });
        Uygula(kalem, sonuc);
    }
}

File: SatinalmaPro.Android/app/src/main/java/com/satinalmapro/android/core/roles/KalemFirmaAtamaYardimcisi.kt
package com.satinalmapro.android.core.roles

import com.satinalmapro.android.core.model.KalemFirmaAtamasi
import com.satinalmapro.android.core.model.TalepKalem
import com.satinalmapro.android.core.model.TeklifItem
import kotlin.math.abs

/** Kalem miktarını firmalara bölme — masaüstü/Shared ile aynı kurallar. */
object KalemFirmaAtamaYardimcisi {
    const val TOLERANS = 0.0001

    fun onayliMi(kalem: TalepKalem): Boolean = etkinAtamalar(kalem).isNotEmpty()

    fun etkinAtamalar(kalem: TalepKalem): List<KalemFirmaAtamasi> {
        val liste = kalem.firmaAtamalari
            .filter { it.teklifId.isNotBlank() && it.miktar > TOLERANS }
            .sortedByDescending { it.miktar }
        if (liste.isNotEmpty()) return liste
        val tid = kalem.onaylananTeklifId?.takeIf { it.isNotBlank() } ?: return emptyList()
        return listOf(
            KalemFirmaAtamasi(
                teklifId = tid,
                miktar = kalem.miktar,
                kabulEdilenMiktar = kalem.kabulEdilenMiktar,
                siparisTamamlandi = kalem.siparisTamamlandi
            )
        )
    }

    fun dogrula(kalem: TalepKalem, atamalar: List<KalemFirmaAtamasi>) {
        require(atamalar.isNotEmpty()) { "«${kalem.malzeme}» için en az bir firma ataması gerekli." }
        require(atamalar.all { it.miktar > TOLERANS }) {
            "«${kalem.malzeme}» atama miktarları sıfırdan büyük olmalı."
        }
        require(atamalar.map { it.teklifId.lowercase() }.distinct().size == atamalar.size) {
            "«${kalem.malzeme}» için aynı firma birden fazla kez seçilemez."
        }
        val toplam = atamalar.sumOf { it.miktar }
        require(abs(toplam - kalem.miktar) <= TOLERANS) {
            "«${kalem.malzeme}» atama toplamı (${"%.2f".format(toplam)}) talep miktarına (${"%.2f".format(kalem.miktar)}) eşit olmalı."
        }
    }

    fun uygula(kalem: TalepKalem, atamalar: List<KalemFirmaAtamasi>): TalepKalem {
        val liste = atamalar
            .filter { it.teklifId.isNotBlank() && it.miktar > TOLERANS }
            .map { it.copy() }
        dogrula(kalem, liste)
        val ana = liste.maxByOrNull { it.miktar }!!.teklifId
        return kabulOzetiniSenkronla(
            kalem.copy(firmaAtamalari = liste, onaylananTeklifId = ana)
        )
    }

    fun tekFirmayaAta(kalem: TalepKalem, teklifId: String): TalepKalem =
        uygula(kalem, listOf(KalemFirmaAtamasi(teklifId = teklifId, miktar = kalem.miktar)))

    fun temizle(kalem: TalepKalem): TalepKalem = kalem.copy(
        firmaAtamalari = emptyList(),
        onaylananTeklifId = null,
        kabulEdilenMiktar = 0.0,
        siparisTamamlandi = false
    )

    fun kabulEkle(kalem: TalepKalem, teklifId: String, miktar: Double): TalepKalem {
        val atamalar = etkinAtamalar(kalem).map { it.copy() }.toMutableList()
        val idx = atamalar.indexOfFirst { it.teklifId.equals(teklifId, true) }
        require(idx >= 0) { "Firma ataması bulunamadı." }
        var atama = atamalar[idx]
        var yeniKabul = atama.kabulEdilenMiktar + miktar
        var yeniMiktar = atama.miktar
        if (yeniKabul > yeniMiktar + TOLERANS) yeniMiktar = yeniKabul
        val tamam = yeniKabul >= yeniMiktar - TOLERANS
        atamalar[idx] = atama.copy(
            miktar = yeniMiktar,
            kabulEdilenMiktar = yeniKabul,
            siparisTamamlandi = tamam
        )
        val ana = kalem.onaylananTeklifId?.takeIf { it.isNotBlank() }
            ?: atamalar.maxByOrNull { it.miktar }!!.teklifId
        return kabulOzetiniSenkronla(kalem.copy(firmaAtamalari = atamalar, onaylananTeklifId = ana))
    }

    fun sevkiyatiTamamla(kalem: TalepKalem, teklifId: String): TalepKalem {
        val atamalar = etkinAtamalar(kalem).map { it.copy() }.toMutableList()
        val idx = atamalar.indexOfFirst { it.teklifId.equals(teklifId, true) }
        require(idx >= 0) { "Firma ataması bulunamadı." }
        var atama = atamalar[idx]
        if (atama.kabulEdilenMiktar < atama.miktar - TOLERANS) {
            atama = atama.copy(miktar = atama.kabulEdilenMiktar)
        }
        atamalar[idx] = atama.copy(siparisTamamlandi = true)
        return kabulOzetiniSenkronla(kalem.copy(firmaAtamalari = atamalar))
    }

    fun kabulOzetiniSenkronla(kalem: TalepKalem): TalepKalem {
        val atamalar = kalem.firmaAtamalari.filter { it.teklifId.isNotBlank() && it.miktar > TOLERANS }
        if (atamalar.isEmpty()) return kalem
        val kabul = atamalar.sumOf { it.kabulEdilenMiktar }
        val tamam = atamalar.all {
            it.siparisTamamlandi || it.kabulEdilenMiktar >= it.miktar - TOLERANS
        }
        val atamaToplam = atamalar.sumOf { it.miktar }
        return kalem.copy(
            kabulEdilenMiktar = kabul,
            siparisTamamlandi = tamam,
            miktar = if (atamaToplam > TOLERANS && abs(atamaToplam - kalem.miktar) > TOLERANS)
                atamaToplam else kalem.miktar
        )
    }

    fun ozetMetni(kalem: TalepKalem, teklifler: List<TeklifItem>): String {
        val atamalar = etkinAtamalar(kalem)
        if (atamalar.isEmpty()) return "Firma seçilmedi"
        return atamalar.joinToString(" + ") { a ->
            val firma = teklifler.firstOrNull { it.id.equals(a.teklifId, true) }?.firmaAdi ?: "?"
            "${"%.2f".format(a.miktar)} $firma"
        }
    }

    /** Firma miktarını ayarlar; diğer firmaların payını kalan miktara oransal ölçekler. */
    fun firmaMiktariniAyarla(kalem: TalepKalem, teklifId: String, miktar: Double): TalepKalem {
        require(miktar > TOLERANS) { "Miktar sıfırdan büyük olmalı." }
        require(miktar <= kalem.miktar + TOLERANS) { "Miktar talep miktarını aşamaz." }
        val diger = etkinAtamalar(kalem)
            .filter { !it.teklifId.equals(teklifId, true) }
            .map { it.copy() }
        val kalan = kalem.miktar - miktar
        if (diger.isEmpty()) {
            require(abs(miktar - kalem.miktar) <= TOLERANS) {
                "Tek firmada miktar talep toplamına eşit olmalı. Başka firma ekleyin veya tüm miktarı seçin."
            }
            return uygula(kalem, listOf(KalemFirmaAtamasi(teklifId = teklifId, miktar = miktar)))
        }
        val sonuc = mutableListOf<KalemFirmaAtamasi>()
        if (kalan > TOLERANS) {
            val digerToplam = diger.sumOf { it.miktar }
            if (digerToplam <= TOLERANS) {
                sonuc += diger.first().copy(miktar = kalan)
            } else {
                for (a in diger) {
                    val m = a.miktar / digerToplam * kalan
                    if (m > TOLERANS) sonuc += a.copy(miktar = m)
                }
                val fark = kalan - sonuc.sumOf { it.miktar }
                if (sonuc.isNotEmpty()) {
                    sonuc[0] = sonuc[0].copy(miktar = sonuc[0].miktar + fark)
                }
            }
        }
        sonuc += KalemFirmaAtamasi(teklifId = teklifId, miktar = miktar)
        return uygula(kalem, sonuc)
    }
}
